Add natural transformation synonym & operator

garyb opened the pull request:
Resolves #23 

Once this is merged `purescript-free` will need an update, as currently that library provides a module `Data.NaturalTransformation`.

garyb committed:
Add natural transformation synonym & operator

garyb commented on src/Data/NaturalTransformation.purs:
```diff
@@ -0,0 +1,14 @@
+-- | instead.
+type NaturalTransformation f g = forall a. f a -> g a
+
+infixr 4 type NaturalTransformation as ~>
```
I figured this gives us enough room to slide `->` underneath while still not making it the lowest precedence, when we decide what that should be.

paf31 replied:
👍 Sounds good.

paf31 commented on src/Data/NaturalTransformation.purs:
```diff
@@ -0,0 +1,14 @@
+module Data.NaturalTransformation where
+
+-- | A type for natural transformations between functors (although the `Functor`
```
I wouldn't mention `Functor` here, and just say "between type constructors" instead. `runCoyoneda` doesn't need this to be a `Functor`, for example.

paf31 commented on src/Data/NaturalTransformation.purs:
```diff
@@ -0,0 +1,14 @@
+-- | A type for natural transformations between functors (although the `Functor`
+-- | constraint is not enforced).
+-- |
+-- | A natural transformation allows a functor to be transformed into another
```
I'm not sure about the connection with `map`. I'd say something like "transforms a functor" and something about uniformity, and give an example like `listToMaybe`, which I think provides a decent intuition.

garyb commented:
Updated the description. I used `fromFoldable` as an example as we don't have `listToMaybe`.

I do think it's worth mentioning that our ~> does not require functor instances though, since technically it should if we're going to call it a natural transformation. I moved it to a note at the end rather than being part of the up front explanation instead though.

hdgarrood commented:
Does parametricity already ensure that, for a natural transformation `f`, we have `f >>> map g == map g >>> f`?

texastoland commented:
Is it not enforced because you might want a `NaturalTransformation` without that constraint or something else?

paf31 commented:
👍 Looks good to me, thanks!

garyb commented:
@AppShipIt yeah, there are lots of things where the constraint would be unhelpful, Phil's coyoneda example being one. 

`Coyoneda f` gives you a `Functor` for any `f`, regardless of whether it has a `Functor` instance or not, so it would not be possible to write the type of `lowerCoyenda` as `Coyoneda f ~> f` without imposing `Functor f`, which kinda defeats the point.

I could be wrong here, so please someone jump in if my ignorance is showing, but I think any type of kind `* -> *` can be considered a lowercase-f-functor regardless of whether there is a `Functor` instance, because our `Functor` is more restrictive than that, as it's for endofunctors.

garyb commented:
Hmm, maybe not, as this is the natural transformation of endofunctors too, the category is fixed as `(->)`.

I don't know :wink:

texastoland commented:
Thanks @garyb 🙏
